datalad/support/parallel.py

```diff
@@ -0,0 +1,572 @@
+        if shutdown:
+            shutdown(force=True)
+
+    def shutdown(self, force=False, exception=None):
+        if self._producer_thread and self._producer_thread.is_alive():
+            # we will try to let the worker to finish "gracefully"
+            self._producer_interrupt = f"shutdown due to {exception}"
+
+        # purge producer queue
+        if self._producer_queue:
+            while not self._producer_queue.empty():
+                self._producer_queue.get()
+
+        lgr.debug("Shutting down %s with %d futures. Reason: %s",
+                  self._executor, len(self._futures), exception)
+
+        if not force:
+            # pop not yet running or done futures.
+            # Those would still have a chance to yield results and finish gracefully
+            # or their exceptions to be bubbled up FWIW.
+            ntotal = len(self._futures)
+            ncanceled = 0
+            nrunning = 0
+            # Do in reverse order so if any job still manages
+            # to sneak in, it would be the earlier submitted one.
+            for k, future in list(self._futures.items())[::-1]:
+                running = future.running()
+                nrunning += int(running)
+                if not (running or future.done()):
+                    if self._futures.pop(k).cancel():
+                        ncanceled += 1
+            lgr.info("Canceled %d out of %d jobs. %d left running.",
+                     ncanceled, ntotal, nrunning)
+        else:
+            # just pop all entirely
+            for k in list(self._futures)[::-1]:
+                self._futures.pop(k).cancel()
+            if self._executor:
+                self._executor.shutdown()
+                self._executor = None
+            if exception:
+                raise exception
+        lgr.debug("Finished shutdown with force=%s due to exception=%r", force, exception)
+
+    def _update_total(self, value):
+        if self.agg:
+            self.total = (
+                self.agg(value, self.total) if self.total is not None else self.agg(value)
+            )
+        else:
+            self.total += 1
+
+    def __iter__(self):
+        jobs = self.jobs
+        if jobs in (None, "auto"):
+            from datalad import cfg
+            # ATM there is no "auto" for this operation, so in both auto and None
+            # just consult max-jobs which can only be an int ATM.
+            # "auto" could be for some auto-scaling based on a single future time
+            # to complete, scaling up/down. Ten config variable could accept "auto" as well
+            jobs = cfg.obtain('datalad.runtime.max-jobs')
+        if jobs >= 1 and not self._can_use_threads:
+            (lgr.debug if ProducerConsumer._alerted_already else lgr.warning)(
+                "Got jobs=%d but we cannot use threads with Pythons versions prior 3.8.0. "
+                "Will run serially", jobs)
+            ProducerConsumer._alerted_already = True
+            jobs = 0
```

Comment:
I think this should be reconsidered.  A caller on python 3.6 or 3.7 that hasn't configured `datalad.runtime.max-jobs` would see a warning every time they simply save:

```console
$ touch foo
$ datalad save
[WARNING] Got jobs=1 but we cannot use threads with Pythons versions prior 3.8.0. Will run serially 
add(ok): /tmp/dl-AIjxceu/foo (file)                                                     
save(ok): /tmp/dl-AIjxceu (dataset)                                                     
action summary:                                                                         
  add (ok: 1)
  save (ok: 1)
```

Reply:
hm, why it got `1`, if I got it right (didn't try) - should have been `"auto"`.  Anyways, I will tune up, so warning is announced only if `jobs > 1`, although setting to explicit `0` even for `1` (at large - the same to the user ;))

Reply:
> hm, why it got `1`, if I got it right (didn't try) - should have been `"auto"`

In the above case, `self.jobs` is None, and when it's None or "auto", it gets set to the value of `datalad.runtime.max-jobs`, which is 1 by default:

https://github.com/datalad/datalad/blob/0ec9383427149cd1e25d09ca5614e21cda0f80ac/datalad/support/parallel.py#L248-L257

<details>
<summary>pdb</summary>

```
(py38-venv) $ datalad save
Total: 0.00 datasets [00:00, ? datasets/s]> /home/kyle/src/python/datalad/datalad/support/parallel.py(250)__iter__()
-> jobs = self.jobs
(Pdb) n
> /home/kyle/src/python/datalad/datalad/support/parallel.py(251)__iter__()
-> if jobs in (None, "auto"):
(Pdb) p jobs
None
(Pdb) where
  /tmp/py38-venv/bin/datalad(33)<module>()
-> sys.exit(load_entry_point('datalad', 'console_scripts', 'datalad')())
  /home/kyle/src/python/datalad/datalad/cmdline/main.py(503)main()
-> ret = cmdlineargs.func(cmdlineargs)
  /home/kyle/src/python/datalad/datalad/interface/base.py(786)call_from_parser()
-> ret = list(ret)
  /home/kyle/src/python/datalad/datalad/interface/utils.py(388)generator_func()
-> for r in _process_results(
  /home/kyle/src/python/datalad/datalad/interface/utils.py(539)_process_results()
-> for res in results:
  /home/kyle/src/python/datalad/datalad/core/local/save.py(353)__call__()
-> yield from ProducerConsumerProgressLog(
  /home/kyle/src/python/datalad/datalad/support/parallel.py(532)__iter__()
-> for res in super().__iter__():
> /home/kyle/src/python/datalad/datalad/support/parallel.py(251)__iter__()
-> if jobs in (None, "auto"):
(Pdb)
```

</details>

Reply:
AH, gotcha! Thank you @kyleam for digesting it for me - very obvious now ;)

Reply:
I made it to issue warning only if above 1 now, so marking it as resolved